File: frontend/src/custom/modules/documentPreview/policy.ts
import {
  normalizePreviewFileType,
  resolveDocumentPreviewType,
  type DocumentPreviewType,
} from '../../../utils/documentPreview.ts'

export type OriginalPreviewMode = 'original' | 'paged_chunks'

export interface KnowledgePreviewPolicy {
  mode: OriginalPreviewMode
  reason?: string
  file_type: string
  file_size: number
  chunk_count: number
  max_original_bytes: number
}

export interface PreviewAdmission {
  mode: OriginalPreviewMode
  reason: string
  maxOriginalBytes: number
}

export const MAX_TEXT_RENDER_BYTES = 512 * 1024
export const DEFAULT_LOADER_BLOB_LIMIT = 8 * 1024 * 1024
export const MAX_COMPLEX_DOCUMENT_CHUNKS = 240

const chunkBoundedTypes = new Set<DocumentPreviewType>(['docx', 'excel', 'pptx'])

export function unwrapKnowledgePreviewPolicy(payload: any): KnowledgePreviewPolicy {
  const value = payload?.data ?? payload
  return {
    mode: value?.mode === 'original' ? 'original' : 'paged_chunks',
    reason: String(value?.reason || ''),
    file_type: normalizePreviewFileType(value?.file_type),
    file_size: Number(value?.file_size || 0),
    chunk_count: Number(value?.chunk_count ?? -1),
    max_original_bytes: Number(value?.max_original_bytes || 0),
  }
}

export function evaluatePreviewAdmission(
  policy: KnowledgePreviewPolicy,
  options: {
    fileType?: string
    fileSize?: number | string
    chunkCount?: number | string
  } = {},
): PreviewAdmission {
  const fileType = normalizePreviewFileType(options.fileType || policy.file_type)
  const previewType = resolveDocumentPreviewType(fileType)
  const observedSize = Number(options.fileSize || policy.file_size || 0)
  const chunkCount = Number(options.chunkCount || policy.chunk_count || 0)
  const limit = Number(policy.max_original_bytes || 0)

  if (policy.mode !== 'original') {
    return {
      mode: 'paged_chunks',
      reason: policy.reason || 'server_policy',
      maxOriginalBytes: limit,
    }
  }
  if (limit <= 0 || observedSize <= 0 || observedSize > limit) {
    return {
      mode: 'paged_chunks',
      reason: observedSize > limit && limit > 0 ? 'file_too_large' : 'unknown_size',
      maxOriginalBytes: limit,
    }
  }
  if (
    chunkBoundedTypes.has(previewType)
    && chunkCount > MAX_COMPLEX_DOCUMENT_CHUNKS
  ) {
    return {
      mode: 'paged_chunks',
      reason: 'too_many_chunks',
      maxOriginalBytes: limit,
    }
  }
  return { mode: 'original', reason: '', maxOriginalBytes: limit }
}

export function blobExceedsAdmission(
  blob: Blob,
  maxOriginalBytes = DEFAULT_LOADER_BLOB_LIMIT,
): boolean {
  const limit = maxOriginalBytes > 0 ? maxOriginalBytes : DEFAULT_LOADER_BLOB_LIMIT
  return blob.size <= 0 || blob.size > limit
}

export function boundedTextBlob(blob: Blob): {
  blob: Blob
  truncated: boolean
} {
  if (blob.size <= MAX_TEXT_RENDER_BYTES) {
    return { blob, truncated: false }
  }
  return {
    blob: blob.slice(0, MAX_TEXT_RENDER_BYTES, blob.type),
    truncated: true,
  }
}

File: internal/custom/modules/documentpreview/policy.go
package documentpreview

import (
	"path/filepath"
	"strings"
)

const (
	ModeOriginal    = "original"
	ModePagedChunks = "paged_chunks"

	MetadataModeKey   = "_system_preview_mode"
	MetadataReasonKey = "_system_preview_reason"

	MaxComplexDocumentChunks int64 = 240
)

const mebibyte int64 = 1024 * 1024

// Policy is the server-authoritative admission result for an in-browser
// original-file preview. A paged_chunks result means the UI must use already
// persisted parser chunks and must not fetch the original object into a Blob.
type Policy struct {
	Mode             string `json:"mode"`
	Reason           string `json:"reason,omitempty"`
	FileType         string `json:"file_type"`
	FileSize         int64  `json:"file_size"`
	ChunkCount       int64  `json:"chunk_count"`
	MaxOriginalBytes int64  `json:"max_original_bytes"`
}

var originalByteLimits = map[string]int64{
	"pdf":      24 * mebibyte,
	"docx":     4 * mebibyte,
	"pptx":     8 * mebibyte,
	"xlsx":     2 * mebibyte,
	"csv":      2 * mebibyte,
	"jpg":      5 * mebibyte,
	"jpeg":     5 * mebibyte,
	"png":      5 * mebibyte,
	"gif":      1 * mebibyte,
	"txt":      2 * mebibyte,
	"text":     2 * mebibyte,
	"md":       2 * mebibyte,
	"markdown": 2 * mebibyte,
	"json":     1 * mebibyte,
	"xml":      1 * mebibyte,
	"html":     1 * mebibyte,
	"css":      1 * mebibyte,
	"js":       1 * mebibyte,
	"ts":       1 * mebibyte,
	"py":       1 * mebibyte,
	"go":       1 * mebibyte,
	"java":     1 * mebibyte,
	"yaml":     1 * mebibyte,
	"yml":      1 * mebibyte,
	"sh":       1 * mebibyte,
	"ini":      1 * mebibyte,
	"conf":     1 * mebibyte,
	"log":      1 * mebibyte,
	"sql":      1 * mebibyte,
	"rs":       1 * mebibyte,
	"rb":       1 * mebibyte,
	"php":      1 * mebibyte,
	"swift":    1 * mebibyte,
	"kt":       1 * mebibyte,
	"scala":    1 * mebibyte,
	"r":        1 * mebibyte,
	"lua":      1 * mebibyte,
	"pl":       1 * mebibyte,
	"toml":     1 * mebibyte,
	"mp3":      20 * mebibyte,
	"wav":      20 * mebibyte,
	"m4a":      20 * mebibyte,
	"flac":     20 * mebibyte,
	"ogg":      20 * mebibyte,
}

var structurallyVerifiedTypes = map[string]struct{}{
	"docx": {},
	"xlsx": {},
	"pptx": {},
	"csv":  {},
}

// Types whose browser decoders cannot currently be bounded by the upload
// structural preflight are intentionally shown through parser chunks.
var alwaysPagedTypes = map[string]struct{}{
	"doc":  {},
	"xls":  {},
	"ppt":  {},
	"bmp":  {},
	"webp": {},
	"tiff": {},
	"svg":  {},
}

func NormalizeFileType(value string) string {
	value = strings.TrimSpace(strings.ToLower(strings.TrimPrefix(value, ".")))
	if value == "" {
		return ""
	}
	if ext := strings.TrimPrefix(strings.ToLower(filepath.Ext(value)), "."); ext != "" {
		return ext
	}
	return value
}

// AnnotateUploadMetadata records the result of the upload-time structural
// preflight. Reserved keys always override caller metadata, preventing a
// request from opting a heavy archive back into an unsafe browser render.
func AnnotateUploadMetadata(
	input map[string]string,
	fileType string,
	heavy bool,
	splitRequired bool,
) map[string]string {
	result := make(map[string]string, len(input)+2)
	for key, value := range input {
		result[key] = value
	}
	normalized := NormalizeFileType(fileType)
	switch {
	case splitRequired:
		result[MetadataModeKey] = ModePagedChunks
		result[MetadataReasonKey] = "physical_split"
	case heavy:
		result[MetadataModeKey] = ModePagedChunks
		result[MetadataReasonKey] = "heavy_structure"
	default:
		if _, alwaysPaged := alwaysPagedTypes[normalized]; alwaysPaged {
			result[MetadataModeKey] = ModePagedChunks
			result[MetadataReasonKey] = "unbounded_decoder"
		} else {
			result[MetadataModeKey] = ModeOriginal
			result[MetadataReasonKey] = ""
		}
	}
	return result
}

func Decide(fileType string, fileSize, chunkCount int64, metadata map[string]string) Policy {
	normalized := NormalizeFileType(fileType)
	limit, supported := originalByteLimits[normalized]
	result := Policy{
		Mode:             ModePagedChunks,
		FileType:         normalized,
		FileSize:         fileSize,
		ChunkCount:       chunkCount,
		MaxOriginalBytes: limit,
	}
	if _, alwaysPaged := alwaysPagedTypes[normalized]; alwaysPaged {
		result.Reason = "unbounded_decoder"
		return result
	}
	if !supported {
		result.Reason = "unsupported"
		return result
	}
	if fileSize <= 0 {
		result.Reason = "unknown_size"
		return result
	}
	if fileSize > limit {
		result.Reason = "file_too_large"
		return result
	}
	if metadata != nil && metadata[MetadataModeKey] == ModePagedChunks {
		result.Reason = strings.TrimSpace(metadata[MetadataReasonKey])
		if result.Reason == "" {
			result.Reason = "upload_preflight"
		}
		return result
	}
	if _, needsVerification := structurallyVerifiedTypes[normalized]; needsVerification {
		if chunkCount < 0 {
			result.Reason = "chunk_count_unavailable"
			return result
		}
		if chunkCount > MaxComplexDocumentChunks {
			result.Reason = "too_many_chunks"
			return result
		}
		if metadata == nil || metadata[MetadataModeKey] != ModeOriginal {
			result.Reason = "unverified_structure"
			return result
		}
	}
	result.Mode = ModeOriginal
	result.Reason = ""
	return result
}

func AllowsOriginal(policy Policy) bool {
	return policy.Mode == ModeOriginal && policy.FileSize > 0 &&
		policy.MaxOriginalBytes > 0 && policy.FileSize <= policy.MaxOriginalBytes
}
